teps/0097-breakpoints-for-taskruns-and-pipelineruns.md

```diff
@@ -0,0 +1,317 @@
+
+On failure of a step, the `-post_file` is written with appending `.err` to it denoting that the previous step has failed with
+and error. The subsequent steps are skipped in this case as well, marking the TaskRun as a failure.
+
+##### Halting a Step on failure
+
+The failed step writes `out.err` to `/tekton/run/<step-no.>` and stops running completely. To be able to debug a step we would
+need it to continue running (not exit), not skip the next steps and signal health of the step. By disabling step skipping, 
+stopping write of the `out.err` file and waiting on a signal by the user to disable the halt, we would be simulating a 
+"breakpoint".
+
+In this breakpoint, which is essentially a limbo state the TaskRun finds itself in, the user can interact with the step 
+environment using a CLI or an IDE. 
+
+##### Exiting breakpoint
+
+To exit a step which has been paused upon failure, the step would wait on a file similar to `<step-no>.breakpointexit` which 
+would unpause and exit the step container. eg: Step 0 fails and is paused. Writing `out.breakpointexit` in `/tekton/run/0/`
+would unpause and exit the step container.
+
+### Debug Environment Additions
+
+#### Mounts
+
+`/tekton/debug/scripts` : Contains scripts which the user can run to mark the step as a success, failure or exit the breakpoint.
+Shared between all the containers.
+
+`/tekton/debug/info/<n>` : Contains information about the step. Single EmptyDir shared between all step containers, but renamed 
+to reflect step number. eg: Step 0 will have `/tekton/debug/info/0`, Step 1 will have `/tekton/debug/info/1` etc.
+
+#### Debug Scripts
+
+`/tekton/debug/scripts/debug-continue` : Mark the step as completed with success by writing to `/tekton/run/<step-no.>/`. eg: User wants to mark
+failed step 0 as a success. Running this script would create `/tekton/run/0/out`.
+
+`/tekton/debug/scripts/debug-continue-failure` : Mark the step as completed with failure by writing to `/tekton/run/<step-no.>/`. eg: User wants to mark
+failed step 0 as a success. Running this script would create `/tekton/run/0/out.err`.
+
+`/tekton/debug/scripts/debug-breakpointexit` : Mark the step as completed with failure by writing to `/tekton/run/<step-no.>/`. eg: User wants to exit
+breakpoint for failed step 0. Running this script would create `/tekton/run/0/out` and `/tekton/run/0/out.breakpointexit`.
+
+## PipelineRun Proposal
+
+Breakpoints would be enabled on PipelineRuns in a way that the logic of putting the breakpoints in place will be in the 
+hands of the TaskRun controller and the step entrypoint.
+
+Using the below spec we will look at how this will work with the following example.
+    
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: quarante-deux-deux
+spec:
+  debug:
+    breakpoint:
+        onFailure: true
+        beforeTask: ["task-build-archive"]
+        afterTask: ["deploy-frontend"]
+```
+
+### Breakpoint before or after a TaskRun
```

Comment:
Nice! :)

Since a pipeline may be running several tasks in parallel, in case of breakpoints, multiple ones could be hit in parallel, and the user will have the opportunity of interacting with all the tasks under debug.
Tasks that do not depend on any task being debugged will continue to run and be scheduled normally, correct?

Reply:
Task dependence is not something I have added right now. I will just update it now.

Reply:
> Tasks that do not depend on any task being debugged will continue to run and be scheduled normally, correct?
Yes. Those tasks will continue running normally.

Reply:
I have added your comment as a statement under `Task Dependence` in this proposal.